lib/Connection.js

```diff
@@ -16,82 +16,71 @@
-    this.conn.I_TRANSPORT_CTL = '*here';
-    this.conn.I_TRANSPORT_IPC = '*NA';
-
-    if (typeof db === 'object') {
-      this.conn.I_TRANSPORT = I_TRANSPORT_DB2;
-      return;
-    }
-
-    if (option && typeof option === 'object') {
-      if (option.host) {
-        this.conn.I_TRANSPORT = I_TRANSPORT_REST;
-        this.conn.I_TRANSPORT_REST_HOST = option.host;
-        if (option.port) {
-          this.conn.I_TRANSPORT_REST_PORT = option.port;
-        } else {
-          this.conn.I_TRANSPORT_REST_PORT = 80;
-        }
-        if (option.path) {
-          this.conn.I_TRANSPORT_REST_PATH = option.path;
-        } else {
-          this.conn.I_TRANSPORT_REST_PATH = '/';
-        }
-      } else {
-        this.conn.I_TRANSPORT = I_TRANSPORT_DB2;
-        if (option.xslib && option.xslib.length > 0) {
-          this.conn.I_XML_SERVICE_LIB = option.xslib;
+  constructor(optionsOrDb, username, password, restOptions) {
+    this.commandList = [];
+    this.returnError = true;
+    this.verbose = false;
+
+    let options = optionsOrDb;
+    // maintain compatibility with pre 1.0 version
+    if (typeof options !== 'object') {
+      options = {};
+      options.database = optionsOrDb;
+      options.username = username;
+      options.password = password;
+      options.transport = 'idb';
+      options.returnError = false;
+
+      if (restOptions && typeof restOptions === 'object') {
+        if (restOptions.host) {
+          // pre v1.0 falls back to idb transport when host is not specified
+          options.transport = 'rest';
         }
+        options = { ...options, ...restOptions };
       }
+    }
 
-      if (option.ctl && option.ctl.length > 0) {
-        this.conn.I_TRANSPORT_CTL = option.ctl;
-      }
-      if (option.ipc && option.ipc.length > 0) {
-        this.conn.I_TRANSPORT_IPC = option.ipc;
-      }
-      if (Number.isNaN(option.buf)) {
-        this.conn.I_TRANSPORT_REST_XML_OUT_SIZE = option.buf.toString();
-      } else {
-        this.conn.I_TRANSPORT_REST_XML_OUT_SIZE = '500000';
-      }
-    } else { this.conn.I_TRANSPORT = I_TRANSPORT_DB2; }
-  }
+    this.transport = options.transport;
 
-  /**
-   * @description override the default timeout value for sync mode.
-   * @param {number} seconds
-   */
-  setTimeout(seconds) {
-    if (typeof seconds === 'number') {
-      this.timeout = seconds * 1000;
+    if (!(this.transport in availableTransports)) {
+      throw new Error(`${this.transport} is not a valid transport`);
     }
+
+    this.transportCall = availableTransports[this.transport];
+
+    if (typeof options.returnError === 'boolean') {
+      this.returnError = options.returnError;
+    }
+
+    if (typeof options.verbose === 'boolean') {
+      this.verbose = options.verbose;
+    }
+    // delete options transports do not need to know
+    delete options.transport;
+    delete options.returnError;
+
+    this.transportOptions = { ...options };
```

Comment:
Why is the user not passing in a `transportOptions` sub object?

Reply:
I was thinking the user would pass in the object in the format like:
```javascript
{
 returnError,
 transport,
 username,
 password,
....
}
```

We could design it for the user to pass in the object like:

```javascript
{
   returnError,
   transport,
   transportOptions: { username, password, ....}
}
```

Reply:
I think that's best. Any transport-specific options are in the `transportOptions` key and you just pass it through to the transport. Then there's no confusion about what key is used for what.